task/host-sp-comms/src/main.rs

```diff
@@ -0,0 +1,314 @@
+task_slot!(SYS, sys);
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum UartLog {
+    Tx(u8),
+    TxFull,
+    Rx(u8),
+    RxOverrun,
+}
+
+ringbuf!(UartLog, 64, UartLog::Rx(0));
+
+/// Notification mask for USART IRQ; must match configuration in app.toml.
+const USART_IRQ: u32 = 1;
+
+#[export_name = "main"]
+fn main() -> ! {
+    let uart = configure_uart_device();
+    let tx_buf = mutable_statics! {
+        static mut UART_TX_BUF: [u8; MAX_MESSAGE_SIZE] = [0; _];
+    };
+    let rx_buf = claim_uart_rx_buf();
+
+    sys_irq_control(USART_IRQ, true);
+
+    loop {
+        // Wait for uart interrupt; if we haven't enabled tx interrupts, this
+        // blocks until there's data to receive.
+        let _ = sys_recv_closed(&mut [], USART_IRQ, TaskId::KERNEL);
+
+        // Clear any RX overrun errors. If we hit this, we will likely fail to
+        // decode the next message from the host, which will cause us to send a
+        // `DecodeFailure` response.
+        if uart.check_and_clear_rx_overrun() {
+            ringbuf_entry!(UartLog::RxOverrun);
+        }
+
+        let maybe_response = 'response: loop {
+            // Receive until we find a message delimiter. Since we're using
+            // corncobs for framing, we're looking for 0x00.
+            while let Some(byte) = uart.try_rx_pop() {
+                ringbuf_entry!(UartLog::Rx(byte));
+
+                if byte == 0x00 {
+                    let n = process_message(tx_buf, rx_buf.as_mut());
+                    rx_buf.clear();
+                    break 'response Some(&tx_buf[..n]);
+                } else {
+                    if rx_buf.push(byte).is_err() {
+                        // Message overflow - nothing we can do here except
+                        // discard data. We'll drop this byte and wait til we
+                        // see a 0 to respond, at which point our
+                        // deserialization will presumably fail and we'll send
+                        // back an error. Should we record that we overflowed
+                        // here?
+                    }
+                }
+            }
+
+            // RX FIFO is empty and we don't have a complete message.
+            break None;
+        };
+
+        // Spin here until we're able to flush our entire response out to the TX
+        // FIFO. We're not going to attempt to read from the RX FIFO while we're
+        // doing this, because the host isn't supposed to be sending us
+        // pipelined requests.
+        if let Some(unframed_data) = maybe_response {
+            let mut iter = corncobs::encode_iter(unframed_data).peekable();
+
+            uart.enable_tx_fifo_empty_interrupt();
+            while let Some(&b) = iter.peek() {
+                if try_tx_push(&uart, b) {
+                    // Discard the byte we just peeked and successfully inserted
+                    // into the TX FIFO.
+                    iter.next().unwrap_lite();
+                    continue;
```

Comment:
Nit: I'd rather have the "TX fifo is full" handling in an `else` block, instead of using `continue` here